File: JVM/jnetproto-java/src/main/java/jnetproto/java/tools/Conversions.java
package jnetproto.java.tools;

import java.lang.reflect.Array;
import java.lang.reflect.Constructor;
import java.lang.reflect.Executable;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.Arrays;
import java.util.List;

import com.xafero.javaenums.BitFlag;
import com.xafero.javaenums.Enums;

import jnetbase.java.meta.Property;
import jnetbase.java.meta.Reflect;
import jnetbase.java.sys.ArrayX;
import jnetproto.java.api.DataType;
import jnetproto.java.core.DataTypes;
import jnetproto.java.core.DataTypes.IDataType;

public final class Conversions {

    public static Object[] convertFor(Object[] args, Executable method) {
        Type[] genTypes = method.getGenericParameterTypes();
        return convert(genTypes, args);
    }

    public static Object[] convert(Type[] types, Object[] args) {
        for (int i = 0; i < args.length; i++)
            args[i] = convert(types[i], args[i]);
        return args;
    }

    public static Object convert(Type type, Object arg) {
        if (type instanceof ParameterizedType) {
        	ParameterizedType param = (ParameterizedType)type;
            Type baseType = param.getRawType();
            Type[] typeArgs = param.getActualTypeArguments();
            if (BitFlag.class.equals(baseType)) {
                return Enums.castToEnum(arg, (Class<?>) typeArgs[0]);
            }
        }
        if (type instanceof Class) {
        	Class clazz = (Class)type;
            if (clazz.isInstance(arg)) {
                return arg;
            }
            if (clazz.isArray()) {
                Class arrayType = clazz.getComponentType();
                if (Enums.isEnum(arrayType)) {
                    int arraySize = Array.getLength(arg);
                    Object array = Array.newInstance(arrayType, arraySize);
                    for (int i = 0; i < arraySize; i++) {
                        Object item = Array.get(arg, i);
                        Object conv = Enums.castToEnum(item, arrayType);
                        Array.set(array, i, conv);
                    }
                    return array;
                }
            }
            if (Enums.isEnum(clazz)) {
                return Enums.castToEnum(arg, clazz);
            }
        }
        return arg;
    }

    private static boolean isUsable(Constructor creator, Object[] args) {
        return creator.getParameters().length == args.length;
    }

    public static Object fromObjectArray(Type type, Object[] args) {
        try {
            IDataType kind = DataTypes.getKind(type);
            if (kind.Kind() != DataType.Unknown)
            {
                if (kind instanceof DataTypes.ArrayDt && ((DataTypes.ArrayDt)kind).Item().Kind() == DataType.Unknown)
                {
                    Class<?> arrayType = ((Class<?>)type).getComponentType();
                    Object array = ArrayX.asTypedArray(args, arrayType, Conversions::convertRaw);
                    return array;
                }
                if (kind instanceof DataTypes.ListDt && ((DataTypes.ListDt)kind).Item().Kind() == DataType.Unknown)
                {
                    Type listType = ((ParameterizedType)type).getActualTypeArguments()[0];
                    List<?> list = ArrayX.asTypedArrayList(args, listType, Conversions::convertRaw);
                    return list;
                }
            }
            Class[] inputTypes = Arrays.stream(args).map(DataTypes::toClass).toArray(Class[]::new);
            Constructor<Object> creator = Reflect.getConstructor(type, inputTypes);
            if (creator == null || !isUsable(creator, args)) {
                creator = Reflect.getFirstConstructor(type);
                if (creator == null || !isUsable(creator, args)) {
                    List<Property> props = Reflect.getProperties(type);
                    if (props.size() == args.length) {
                        Object obj = Reflect.createNew(type);
                        for (int i = 0; i < props.size(); i++) {
                            Object arg = args[i];
                            Property prop = props.get(i);
                            prop.Set().invoke(obj, arg);
                        }
                        return obj;
                    }
                    throw new IllegalArgumentException("No constructor: " + type);
                }
            }
            Parameter[] outputTypes = creator.getParameters();
            args = Conversions.convertFor(args, creator);
            args = convertFor(args, outputTypes);
            return creator.newInstance(args);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static Object[] convertFor(Object[] args, Parameter[] parameters)
    {
        for (int i = 0; i < args.length && i < parameters.length; i++)
        {
            Type prm = parameters[i].getParameterizedType();
            Object arg = args[i];
            Object value = convertRaw(arg, prm);
            args[i] = value;
        }
        return args;
    }

    private static Object convertRaw(Object v, Type r)
    {
        Class t = r instanceof Class<?> ? (Class)r : (Class) ((ParameterizedType)r).getRawType();
        if (t.isEnum())
            return v;
        if (t.equals(Object.class))
            return v;
        if (t.isInstance(v))
            return v;
        if (v instanceof Object[])
            return fromObjectArray(r, (Object[])v);
        return v;
    }

    public static Object toObjectArray(Object obj) {
        try {
            IDataType kind = DataTypes.getKind(obj);
            if (kind.Kind() != DataType.Unknown) {
                if (kind instanceof DataTypes.ArrayDt && ((DataTypes.ArrayDt)kind).Item().Kind() == DataType.Unknown)
                    obj = ArrayX.asObjectArray(obj);
                if (kind instanceof DataTypes.ListDt && ((DataTypes.ListDt)kind).Item().Kind() == DataType.Unknown)
                    obj = ArrayX.asObjectArray((Iterable)obj);
                if (obj instanceof Object[] && Arrays.stream((Object[])obj)
                        .anyMatch(a -> DataTypes.getKind(a).Kind() == DataType.Unknown)) {
                	Object[] ar = (Object[])obj;
                    for (int i = 0; i < ar.length; i++)
                        ar[i] = toObjectArray(ar[i]);
                }
                return obj;
            }
            Class type = obj.getClass();
            List<Property> props = Reflect.getProperties(type);
            Object[] args = new Object[props.size()];
            for (int i = 0; i < args.length; i++) {
                Property prop = props.get(i);
                Method propGet = prop.Get();
                Object propVal = propGet.invoke(obj);
                args[i] = toObjectArray(propVal);
            }
            return args;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: NET/JNetProto.Sharp/Tools/Conversions.cs
using System;
using System.Collections;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using JNetBase.Sharp.Sys;
using JNetProto.Sharp.API;
using JNetProto.Sharp.Core;

namespace JNetProto.Sharp.Tools
{
    public static class Conversions
    {
        public static object Convert(Type type, object arg)
        {
            if (type.IsInstanceOfType(arg))
            {
                return arg;
            }
            if (arg is ITuple tuple)
            {
                var values = new object[tuple.Length];
                for (var i = 0; i < values.Length; i++)
                    values[i] = tuple[i];
                var conv = Activator.CreateInstance(type, values);
                return conv;
            }
            return arg;
        }

        private static bool IsUsable(ConstructorInfo creator, object[] args)
        {
            return creator.GetParameters().Length == args.Length;
        }

        public static object FromObjectArray(Type type, object[] args)
        {
            var kind = DataTypes.GetKind(type);
            if (kind.Kind != DataType.Unknown)
            {
                if (kind is DataTypes.ArrayDt at && at.Item.Kind == DataType.Unknown)
                {
                    var arrayType = type.GetElementType();
                    var array = Arrays.AsTypedArray(args, arrayType, ConvertRaw);
                    return array;
                }
                if (kind is DataTypes.ListDt lt && lt.Item.Kind == DataType.Unknown)
                {
                    var listType = type.GenericTypeArguments[0];
                    var list = Arrays.AsTypedArrayList(args, listType, ConvertRaw);
                    return list;
                }
            }
            var inputTypes = args.Select(DataTypes.ToType).ToArray();
            var creator = type.GetConstructor(inputTypes);
            if (creator == null || !IsUsable(creator, args))
            {
                creator = type.GetConstructors().FirstOrDefault();
                if (creator == null || !IsUsable(creator, args))
                {
                    var props = type.GetProperties();
                    if (props.Length == args.Length)
                    {
                        var obj = Activator.CreateInstance(type);
                        for (var i = 0; i < props.Length; i++)
                        {
                            var arg = args[i];
                            var prop = props[i];
                            prop.SetValue(obj, arg);
                        }
                        return obj;
                    }
                    throw new ArgumentException($"No constructor: {type}");
                }
            }
            var outputTypes = creator.GetParameters();
            args = ConvertFor(args, outputTypes);
            return creator.Invoke(args);
        }

        private static object[] ConvertFor(object[] args, ParameterInfo[] parameters)
        {
            for (var i = 0; i < args.Length && i < parameters.Length; i++)
            {
                var prm = parameters[i].ParameterType;
                var arg = args[i];
                var value = ConvertRaw(arg, prm);
                args[i] = value;
            }
            return args;
        }

        private static object ConvertRaw(object v, Type t)
        {
            if (t.IsEnum)
                return v;
            if (t == typeof(object))
                return v;
            if (t.IsInstanceOfType(v))
                return v;
            if (v is object[] va)
                return FromObjectArray(t, va);
            return System.Convert.ChangeType(v, t);
        }

        public static object ToObjectArray(object obj)
        {
            var kind = DataTypes.GetKind(obj);
            if (kind.Kind != DataType.Unknown)
            {
                if (kind is DataTypes.ArrayDt at && at.Item.Kind == DataType.Unknown)
                    obj = Arrays.AsObjectArray((IEnumerable)obj);
                if (kind is DataTypes.ListDt lt && lt.Item.Kind == DataType.Unknown)
                    obj = Arrays.AsObjectArray((IEnumerable)obj);
                if (obj is object[] ar && ar.Any(a => DataTypes.GetKind(a).Kind == default))
                    for (var i = 0; i < ar.Length; i++)
                        ar[i] = ToObjectArray(ar[i]);
                return obj;
            }
            var type = obj.GetType();
            var props = type.GetProperties();
            var args = new object[props.Length];
            for (var i = 0; i < args.Length; i++)
                args[i] = ToObjectArray(props[i].GetValue(obj));
            return args;
        }
    }
}